hazelcast/src/main/java/com/hazelcast/datalink/ReferenceCounter.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.datalink;
+
+import com.hazelcast.core.HazelcastException;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ReferenceCounter {
+
+    private final AtomicInteger referenceCount = new AtomicInteger(1);
+    private final Callable<Void> destroy;
+
+    public ReferenceCounter(Callable<Void> destroy) {
+        this.destroy = destroy;
+    }
+
+    public ReferenceCounter retain() {
+        int oldCount = referenceCount.getAndIncrement();
```

Comment:
retain may be called and release almost at the same time - referenceCount will be increased, release will decrease it immediatelly and we end up with negative counter and exception, despite that both retain and release was called when state allowed it to do so.

Reply:
See the new javadoc if you still think this is an issue.